src/main/java/net/neoforged/neoforge/registries/DataMapLoader.java

```diff
@@ -0,0 +1,178 @@
+        });
+        final Map<ResourceKey<R>, T> newMap = new IdentityHashMap<>();
+        result.forEach((key, val) -> newMap.put(key, val.attachment()));
+
+        return newMap;
+    }
+
+    private CompletableFuture<Map<ResourceKey<? extends Registry<?>>, LoadResult<?>>> load(ResourceManager manager, Executor executor, ProfilerFiller profiler) {
+        return CompletableFuture.supplyAsync(() -> load(manager, profiler, registryAccess, conditionContext), executor);
+    }
+
+    private static Map<ResourceKey<? extends Registry<?>>, LoadResult<?>> load(ResourceManager manager, ProfilerFiller profiler, RegistryAccess access, ICondition.IContext context) {
+        final RegistryOps<JsonElement> ops = ConditionalOps.create(RegistryOps.create(JsonOps.INSTANCE, access), context);
+
+        final Map<ResourceKey<? extends Registry<?>>, LoadResult<?>> values = new HashMap<>();
+        access.registries().forEach(registryEntry -> {
+            final var registryKey = registryEntry.key();
+            profiler.push("registry_data_maps/" + registryKey.location() + "/locating");
+            final var fileToId = FileToIdConverter.json(PATH + "/" + getFolderLocation(registryKey.location()));
+            for (Map.Entry<ResourceLocation, List<Resource>> entry : fileToId.listMatchingResourceStacks(manager).entrySet()) {
+                ResourceLocation key = entry.getKey();
+                final ResourceLocation attachmentId = fileToId.fileToId(key);
+                final var attachment = RegistryManager.getDataMap((ResourceKey) registryKey, attachmentId);
+                if (attachment == null) {
+                    LOGGER.warn("Found data map file for non-existent data map type '{}' on registry '{}'.", attachmentId, registryKey.location());
+                    continue;
+                }
+                profiler.popPush("registry_data_maps/" + registryKey.location() + "/" + attachmentId + "/loading");
+                values.computeIfAbsent(registryKey, k -> new LoadResult<>(new HashMap<>())).results.put(attachment, readData(
+                        ops, attachment, (ResourceKey) registryKey, entry.getValue(), context));
+            }
+            profiler.pop();
+        });
+
+        return values;
+    }
+
+    public static String getFolderLocation(ResourceLocation registryId) {
+        return (registryId.getNamespace().equals(ResourceLocation.DEFAULT_NAMESPACE) ? "" : registryId.getNamespace() + "/") + registryId.getPath();
+    }
+
+    private static <A, T> List<DataMapFile<A, T>> readData(RegistryOps<JsonElement> ops, DataMapType<A, T> attachmentType, ResourceKey<Registry<T>> registryKey, List<Resource> resources, ICondition.IContext context) {
+        final var codec = DataMapFile.codec(registryKey, attachmentType);
+        final List<DataMapFile<A, T>> entries = new LinkedList<>();
+        for (final Resource resource : resources) {
+            try {
+                try (Reader reader = resource.openAsReader()) {
```

Comment:
Why the nested try statements? Just make the upper level try statement be the try with resources